### PR TITLE
i#5131 new linux syscalls: handle close_range.

### PR DESCRIPTION
**Pull request by abhinav92003**
Adds handling for the close_range syscall in Linux. For a contiguous range of FDs,
close_range allows either closing them or marking them as close-on-exec. We
convert the seen close_range call into multiple close_range calls, one for each
contiguous sub-range that does not have any DR-private FDs. We always skip
executing the app's close_range itself.

Fixes the file_io test that assumed that there are only 4096 FDs. Instead, we
get the rlim_max for RLIMIT_NOFILE and try to use the DR_STEAL_FDS FDs after it
to test whether DR's handling of close and close_range is working as expected.

Adds tests for close_range to verify correct happy-case behaviour and EINVAL.
close_range does not propagate errors from closing individual FDs; tests that as
well.

Manually sets the SYS_close_range constant to __NR_close_range. This is needed
because the header with which we updated the DR syscall table defines the latter
but not the former. It is still not defined for AArch64 yet, which doesn't even have
the NR constant right now. We need to update the syscall defines in uapi.h from a
newer kernel version; this will be done in a follow up PR.

Issue: #5131

### CODE REVIEW

File: core/unix/os.c [hunk 1]
```diff
@@ -7585,6 +7601,71 @@ pre_system_call(dcontext_t *dcontext)
...
+            set_failure_return_val(dcontext, EINVAL);
+            DODEBUG({ dcontext->expect_last_syscall_to_fail = true; });
+            break;
+        }
+        uint cur_range_first_fd, cur_range_last_fd;
+        bool cur_range_valid = false;
+        int ret = 0;
+        for (int i = first_fd; i <= last_fd; i++) {
+            /* Do not allow any changes to DR-owned FDs. */
+            if ((is_cloexec && fd_is_dr_owned(i)) ||
```
Comment (derekbruening) [inline core/unix/os.c:7631]: If DR fds are all in a single range, it is more performant to let the close_range syscall execute for the portion that does not overlap.  After all, they created this new syscall purely for performance, so maybe we shouldn't undo it.

For the default -steal_fds DR fds are in a single range -- unless we run out I guess: so even w/ default ops there can be DR fds sprinkled about, although it's unlikely.

I would vote for checking for zero overlap, which should be the common case (probably 90% of cases if people use the rlimit as the upper bound) and letting the close_range syscall execute in that case at least.  That's the fastpath; any overlap at all and we go to this slowpath.

Update: I was misled by the PR description which made me think DR turned the app's close_range into a loop of close calls.  I see this is making close_range calls.  So the only action item then is to maybe tweak the description to make this more clear.

File: suite/tests/client-interface/file_io.c [hunk 2]
```diff
@@ -86,8 +86,74 @@ main()
...
+    if (syscall(__NR_close_range, rlimit.rlim_max - 1, rlimit.rlim_max - 1,
+                CLOSE_RANGE_CLOEXEC) == -1) {
+        perror("close_range mark as close-on-exec failed");
+    }
+#        else
+    assert(fcntl(rlimit.rlim_max - 1, F_SETFD,
+                 fcntl(rlimit.rlim_max - 1, F_GETFD) | FD_CLOEXEC) == 0);
+#        endif
+    if (!TEST(FD_CLOEXEC, fcntl(rlimit.rlim_max - 1, F_GETFD)))
+        fprintf(stderr, "clone_range failed to set the close-on-exec flag\n");
```
Comment (derekbruening) [inline suite/tests/client-interface/file_io.c:134]: s/clone/close/

**Review by derekbruening — APPROVED** [review]

File: core/unix/os.c (hunk 1 above)
Comment (abhinav92003) [reply core/unix/os.c:7631]: The description mentions that we use "multiple such calls, one for each contiguous sub-range". I guess I just need to add "close_range" explicitly there.

Comment (abhinav92003) [reply core/unix/os.c:7631]: Done.

File: core/unix/os.c [hunk 3]
```diff
@@ -7585,6 +7601,71 @@ pre_system_call(dcontext_t *dcontext)
...
+        uint flags = sys_param(dcontext, 2);
+        bool is_cloexec = TEST(CLOSE_RANGE_CLOEXEC, flags);
+        if (is_cloexec) {
+            /* client.file_io has a test for CLOSE_RANGE_CLOEXEC, but it hasn't been
+             * verified on a system with kernel version >= 5.11 yet.
+             */
+            ASSERT_NOT_TESTED();
+        }
+        /* We do not let the app execute their own close_range ever. Instead we
+         * we make multiple close_range syscalls ourselves, one for each contiguous
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/os.c:7617]: Done.

File: core/unix/include/close_range.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/****************************************************************************
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/include/close_range.h:1]: Done. Sounded the same in my mind before, but I agree was a bit awkward.

File: suite/tests/client-interface/file_io.c (hunk 2 above)
Comment (abhinav92003) [reply suite/tests/client-interface/file_io.c:134]: Done.

File: core/unix/include/syscall_linux_arm.h [hunk 5]
```diff
@@ -2970,6 +2970,16 @@
...
 #    define SYS_set_tls __ARM_NR_set_tls
+
+/* The following SYS_* constants are defined manually for some of the above
+ * __NR_* constants that do not have a corresponding SYS_* constant defined
+ * in the header files yet. This is so that we can add support for the
+ * corresponding syscall. These entries should be deleted when the above
+ * list is updated with a newer header file that contains them already.
+ */
+#    ifdef __NR_close_range
+#        define SYS_close_range __NR_close_range
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/include/syscall_linux_arm.h:2981]: The Linux kernel that I have on my aarch64 machine doesn't have that syscall yet. It seems pretty old -- 5.4. I downloaded headers for `aarch64-linux-gnu` on my x86 machine (which is 5.10) and it indeed defines it.

I also see that `client.file_io` is not enabled on Linux yet. I'll do both in a follow up PR.

[the remark this answers is not in the document]
Comment (derekbruening) [reply core/unix/include/syscall_linux_arm.h:2981]: > I also see that client.file_io is not enabled on Linux yet. I'll do both in a follow up PR.

You mean aarchxx?

[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/include/syscall_linux_arm.h:2981]: I mean updating the syscalls defines for aarch64 (in uapi.h) from a newer version and enabling the file_io test on aarch64 (not checked whether it's enabled already on arm yet)

[the remark this answers is not in the document]
Comment (abhinav92003) [reply core/unix/include/syscall_linux_arm.h:2981]: > > I also see that client.file_io is not enabled on Linux yet. I'll do both in a follow up PR.
> 
> You mean aarchxx?

Yeah. Sorry for the confusion.

**Pull request merged**